### ISSUE TITLE
Detect funcs implementing defined interface

### ISSUE THREAD

**Issue by mvdan** [golang/lint#218]
If one does `golint go/ast`, the vast majority of the warnings are of the form:

```
exported method *.Pos should have comment or be unexported
exported method *.End should have comment or be unexported
```

All of these methods implement `Node.Pos` and `Node.End`, which do have documentation.

I hit a very similar issue in a package of mine. As a developer, to fix these golint warnings I would basically just repeat the same godoc line for every single implementation without any changes, which doesn't make much sense.

What I suggest is that golint (and perhaps even godoc too?) notice if functions are implementing an exported interface defined in the same package, and reuse its godoc as a fallback if the implementation has none.

This would only work if there is exactly one exported interface that matches the function name and signature in the same package, so if two interfaces were to define `Pos() Pos`, then golint would warn about all its implementations again.

Perhaps in the broader context this does not make sense, as implementations of interfaces might not always want to share the same godoc as a fallback. So maybe this isn't a good general rule. But the current state seems sort of broken to me.

**Comment by savinov** [comment]
I agree: repeating same docs in all implementations is dull.
Maybe this is a godoc issue: to inherit docs from the interface definition to all the implementations.
In Java is is solved quite elegant: you just put {@inheritDoc} tag and you have javadoc from the interface method in an implementation.
That's what godoc doesn't have, but should have I think.

**Comment by mvdan** [comment]
When I said `perhaps even godoc too` in the original issue, I meant that two solutions come to mind:

* `golint` does this change, but `godoc` doesn't. This is assuming that the user understands e.g. what `Pos()` means from reading the interface definition. This is also the less intrusive change, as no godoc pages would change.
* Both change. An advantage is that things would stay consistent and that `godoc` would get this also fixed, if it is seen as an issue rather than something avoided by design.

**Comment by Quasilyte** [comment]
It is especially annoying for ADT-like interfaces.
`go/ast.Node` is one of this kind.
Suppose you have interface `Object` which has dozens of implementations,
each should implement `Type()` method.
Should I **really** write `Type returns underlying Object type` everywhere?

**Comment by smyrman** [comment]
Don't know if this is feasable, but if it hard to detect if a method should not require a doc string during lint because it implements some interface, maybe one could make it explicit by looking for compile-time checks within the same package.

E.g. If one find:

```go
// Dog can Bark.
type Dog interface {
    Bark() string
}

// Compile-time checks.
var (
    _ Dog = Affenpinscher{}
)


// Affenpinscher is a type of dog.
type Affenpinscher struct {}

func (dog Affenpinscher) Bark() string {
    return "Wooof!"
}

**Comment by mvdan** [comment]
Thank you for submitting this issue! As per https://github.com/golang/go/issues/38968, we are freezing and deprecating golint. There's no drop-in replacement to golint per se, but you should find that [Staticcheck](https://staticcheck.io/) works well in encouraging good Go code, much like golint did in the past, since it also includes style checks. There's always `gofmt` and `go vet` too, of course.

If you would like to contribute further, I'd encourage you to engage [Staticcheck's issue tracker](https://github.com/dominikh/go-tools/issues) or look at [vet's open issues](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+cmd%2Fvet+in%3Atitle), as they are both actively maintained. If you have an idea that doesn't fit into either of those tools, you could look at [other Go linters](https://github.com/golangci/awesome-go-linters), or write your own - these days it's fairly straightforward with [go/analysis](https://pkg.go.dev/golang.org/x/tools/go/analysis).

To help avoid confusion, I'm closing all issues before we freeze the repository. If you have any feedback, you can leave a comment on the proposal thread where it was decided to deprecate golint - though note that the proposal has been accepted for nearly a year. Thanks!
